File: Python/encryption.py
import hashlib
from Crypto import Random
from Crypto.Cipher import AES
import hmac
import tkMessageBox

## Constant values

## Size of master key matches size of aes key and hmac key.
MASTER_KEY_SIZE_BYTES = 32
## Most sources agree 32 bytes is long enough for a salt.
DEFAULT_SALT_SIZE_BYTES = 32
## Required block size for AES...cannot be changed.
AES_BLOCK_SIZE_BYTES = 16
## Required size for AES IV...cannot be changed.
AES_IV_SIZE_BYTES = 16
## Set to take approximately 1 second on my 2014 mid-range laptop.
KEY_DERIVATION_ITERATIONS = 100000

## This is called to generate a random byte array.  An output size may
## be specified.  If not specified, output size defaults to
## DEFAULT_SALT_SIZE_BYTES.
def generate_salt(salt_size_bytes = DEFAULT_SALT_SIZE_BYTES):
    return Random.new().read(salt_size_bytes)

## Take in password and salt derive derive the aes key and the hmac key.
def derive_keys(password, salt):
    ## Derive master key using PBKDF2 (slow derivation process).
    master_key = hashlib.pbkdf2_hmac('sha1', password, salt, 
                     KEY_DERIVATION_ITERATIONS, MASTER_KEY_SIZE_BYTES)
    ## Derive aes key and hmac key from master key.
    ## Use HKDF-Expand
    ## https://tools.ietf.org/html/rfc5869
    aes_key_exp_byte = '\x01'
    hmac_key_exp_byte = '\x02'
    aes_key = calculate_mac_tag(master_key, aes_key_exp_byte)
    hmac_key = calculate_mac_tag(master_key, aes_key + hmac_key_exp_byte)
    return aes_key, hmac_key

## Take in hmac key and message and return mac tag.
def calculate_mac_tag(hmac_key, message):
    hmac_function = hmac.new(hmac_key, message, hashlib.sha256)
    return hmac_function.digest()
  
## Take plaintext and password.  Return key salt, aes iv, ciphertext,
## and mac tag.
def encrypt_then_authenticate(plaintext, password):

    ## Generate a new key salt each time we encrypt. Use default salt size.
    key_salt = generate_salt()
    ## Generate a new aes IV each time we encrypt. Must use required IV size.
    aes_iv = generate_salt(AES_IV_SIZE_BYTES)
    ## Populate aes_key and hmac_key via key derivation function.
    aes_key, hmac_key = derive_keys(password, key_salt)
  
    ## In AES CBC mode the plaintext needs have a length that is a multiple
    ## of the AES block size.  First, figure out how many extra bytes we
    ## need to add. 
    padding_length = (AES_BLOCK_SIZE_BYTES - 
                     (len(plaintext) % AES_BLOCK_SIZE_BYTES))
    ## Now, set plaintext_padded equal to the plaintext plus a string
    ## of padding in which the value in each byte is the total length
    ## of padding needed.
    plaintext_padded = plaintext + chr(padding_length) * padding_length
    
    ## Run encryption on padded plaintext to get ciphertext.
    aes_function = AES.new(aes_key, AES.MODE_CBC, aes_iv)  
    ciphertext = aes_function.encrypt(plaintext_padded)

    ## Run hmac to create a message authentication tag (mac).
    mac_tag = calculate_mac_tag(hmac_key, aes_iv + ciphertext) 
    return key_salt, aes_iv, ciphertext, mac_tag

## First, check that encrypted data matches mac tag.  If it does,
## decrypt and return decrypted data.
def authenticate_then_decrypt(ciphertext, password, key_salt, aes_iv, mac_tag):

    ## Populate aes_key and hmac_key via key derivation function.
    aes_key, hmac_key = derive_keys(password, key_salt)
  
    ## Run hmac to create a message authentication tag (mac).
    calculated_mac_tag = calculate_mac_tag(hmac_key, aes_iv + ciphertext) 

    ## Test if calculated mac tag matches expected mac tag.
    ## If it doesn't it means a bad password was entered or
    ## the data is corrupted.  Exit function.
    hmac_matched = hmac.compare_digest(mac_tag, calculated_mac_tag)
    if not hmac_matched:
        tkMessageBox.showerror("Decryption error!", "Incorrect password" +
                               " or corrupted data.")
        return
 
    ## Mac tag check passed, now run decryption routine.
    aes_function = AES.new(aes_key, AES.MODE_CBC, aes_iv)  
    plaintext = aes_function.decrypt(ciphertext)

    ## We need to "unpad" the result (data to be encrypted is padded to be
    ## a length that is a multiple of the AES block size).  After decrypting
    ## we need to get rid of what was added to the end.  This code first
    ## finds the value in the last character of the string, then removes
    ## x last characters where x is the value of the last character in
    ## the string.
    plaintext_unpadded = plaintext[:-ord(plaintext[-1])]
  
    ## Return unpadded plaintext.
    return plaintext_unpadded

File: CSharp/SecurePastebin/Encryption.cs
using System;
using System.IO;
using System.Windows.Forms;
using System.Security.Cryptography;

namespace SecurePastebin
{
    internal static class Encryption
    {
        // Size of master key matches size of aes key and hmac key.
        private const int masterKeySizeBytes = 32;
        // Most sources agree 32 bytes is long enough for a salt.
        private const int defaultSaltSizeBytes = 32;
        // Required size for AES IV...cannot be changed.
        private const int aesIVSizeBytes = 16;
        // Set to take approximately 1 second on my 2014 mid-range laptop.
        private const int keyDerivationIterations = 100000;

        // Take in two arrays and return an array that joins array1 | array2.
        private static byte[] CombineArrays(byte[] array1, byte[] array2)
        {
            var combinedArray = new byte[array1.Length + array2.Length];
            array1.CopyTo(combinedArray, 0);
            array2.CopyTo(combinedArray, array1.Length);
            return combinedArray;
        }

        // This is called to generate a random byte array.  An ouput size may be specified.
        // If not specified, output size defaults to defaultSaltSizeBytes.
        public static byte[] GenerateSalt(int saltSizeBytes = defaultSaltSizeBytes)
        {
            using (var randomGenerator = new RNGCryptoServiceProvider())
            {
                var salt = new byte[saltSizeBytes];
                randomGenerator.GetBytes(salt);
                return salt;
            }
        }

        // Take in password and salt and derive the aes key and the hmac key.
        private static void DeriveKeys(string password, byte[] salt, out byte[] aesKey, out byte[] hmacKey)
        {
            // Derive master key using PBKDF2 (slow derivation process).
            byte[] masterKey;
            using (var derivedKey = new Rfc2898DeriveBytes(password, salt, keyDerivationIterations))
            {
                masterKey = derivedKey.GetBytes(masterKeySizeBytes);
            }
            // Derive aes key and hmac key from master key.
            // See HKDF-Expand definition in section 2.3 of https://tools.ietf.org/html/rfc5869
            var aesKeyExpansionByte = new byte[] { 0x01 };
            var hmacKeyExpansionByte = new byte[] { 0x02 };
            aesKey = CalculateMacTag(masterKey, aesKeyExpansionByte);
            hmacKey = CalculateMacTag(masterKey, CombineArrays(aesKey, hmacKeyExpansionByte));
        }

        // Take in hmacKey and message and return mac tag.
        private static byte[] CalculateMacTag(byte[] hmacKey, byte[] message)
        {
            using (var hmac = new HMACSHA256(hmacKey))
            {
                return hmac.ComputeHash(message);
            }
        }

        // Take in plaintext and password. Populate output parameters keySalt, aesIV, cipherText, and macTag.
        public static void EncryptThenAuthenicate(string plainText, string password, out byte[] keySalt, out byte[] aesIV, out byte[] cipherText, out byte[] macTag)
        {
            // Default output params to null
            keySalt = null;
            aesIV = null;
            cipherText = null;
            macTag = null;
            try
            {
                // Generate a new key salt each time we encrypt.  Use default salt size.
                keySalt = GenerateSalt();
                // Generate a new aes IV each time we encrypt.  Must use required IV size.
                aesIV = GenerateSalt(aesIVSizeBytes);

                // Populate aesKey and hmacKey via key derivation function.
                byte[] aesKey, hmacKey;
                DeriveKeys(password, keySalt, out aesKey, out hmacKey);

                // Set up AES encryption provider object.
                using (var aesAlg = new AesCryptoServiceProvider())
                {
                    aesAlg.Key = aesKey;
                    aesAlg.IV = aesIV;
                    aesAlg.Mode = CipherMode.CBC;
                    aesAlg.Padding = PaddingMode.PKCS7;

                    // Create an encrytor to perform the stream transform.
                    var encryptor = aesAlg.CreateEncryptor(aesKey, aesIV);

                    // Create the streams used for encryption.
                    using (var msEncrypt = new MemoryStream())
                    {
                        using (var csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
                        {
                            using (var swEncrypt = new StreamWriter(csEncrypt))
                            {
                                // Write all data to the stream.
                                swEncrypt.Write(plainText);
                            }
                            cipherText = msEncrypt.ToArray();
                        }
                    }
                }
                // Calculate mac tag.
                macTag = CalculateMacTag(hmacKey, CombineArrays(aesIV, cipherText));
            }
            catch (Exception ex)
            {
                MessageBox.Show(String.Format("Encryption failed. {0}", ex.Message), "Encryption Error!", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        // Takes in cipherText, password, key salt, aes IV, and mac tag.
        // First, authenticates aesIV||cipherText using mac tag.
        // Then, if that succeeds, decrypts and returns plainText.
        public static string AuthenticateThenDecrypt(byte[] cipherText, string password, byte[] keySalt, byte[] aesIV, byte[] macTag)
        {
            try
            {
                // Populate aesKey and hmacKey via key derivation function.
                byte[] aesKey, hmacKey;
                DeriveKeys(password, keySalt, out aesKey, out hmacKey);

                // Calculate mac tag using hmacKey on aesIV||cipherText
                var calcMacTag = CalculateMacTag(hmacKey, CombineArrays(aesIV, cipherText));

                // Compare to see if stored mac tag matches calculated mac tag.
                // Use constant time comparison to prevent timing attacks.
                var mismatch = 0;
                for (var i = 0; i < macTag.Length; i++)
                    mismatch |= macTag[i] ^ calcMacTag[i];

                // If message doesn't authenticate, return error and exit routine.
                if (mismatch != 0)
                    throw new Exception("Incorrect password or corrupted data.");

                // Authentication suceeded, now decrypt...

                // Set up AES encryption provider object.
                using (var aesAlg = new AesCryptoServiceProvider())
                {
                    aesAlg.Key = aesKey;
                    aesAlg.IV = aesIV;
                    aesAlg.Mode = CipherMode.CBC;
                    aesAlg.Padding = PaddingMode.PKCS7;

                    // Create a decrytor to perform the stream transform.
                    var decryptor = aesAlg.CreateDecryptor(aesKey, aesIV);

                    // Create the streams used for decryption.
                    using (var msDecrypt = new MemoryStream(cipherText))
                    {
                        using (var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Read))
                        {
                            using (var srDecrypt = new StreamReader(csDecrypt))
                            {
                                // Read the decrypted bytes from the decrypting stream and return as string.
                                return srDecrypt.ReadToEnd();
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(String.Format("Decryption failed. {0}", ex.Message), "Decryption Error!", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return string.Empty;
            }
        }
    }
}

